design/pps.md

```diff
@@ -0,0 +1,86 @@
+
+When a pod starts up, it sends a request to PPS to ask for a chunk to process.  Currently, a pod exits once it has processed a chunk.  This works because currently the number of pods is always equal to the number of chunks.  In the future however, we should decouple pods and chunks to enable more advanced features such as straggler mitigation.  For instance, if a chunk is taking too long to process, we can simply split the chunk into smaller chunks so the other pods can share the work.  See [#442](https://github.com/pachyderm/pachyderm/issues/442).
+
+## Monitor the liveness of pods
+
+Normally, if a pod crashes, Kubernetes will instantiate a new pod, due to the fact that a Kubernetes job always ensures that its pods successfully terminate.  However, the new pod will have a completely new identity, and thus when it comes to PPS to ask for a chunk, PPS needs to somehow know to hand it the chunk that was originally being processed by the crashed pod.
+
+Furthermore, it's possible for a pod to get into a state where it's still "alive" from the perspective of Kubernetes (and thus a new pod is not created), but is not making any progress from the perspective of the pipeline.  For instance, there could be a network partition between a pod and PPS/PFS, but no partition between the pod and the Kubernetes master.
+
+To handle these challenges, PPS uses a lease system to manage chunks.  When a pod asks for a chunk, PPS "leases" the chunk to the pod, and starts an internal timer that revokes the lease after a configurable period.  To keep a lease, a pod needs to continuously send heartbeats to PPS to report that it's alive.  Everytime such a heartbeat is received, PPS renews the lease and restarts the timer.
+
+Now let's consider what happens when a pod crashes.  Because the pod crashed, it stops sending heartbeats to PPS, so the lease on the chunk is automatically revoked after a while.  As soon as the lease expires, PPS can give the chunk to the new pod who has been requesting a chunk.  Thus we've completed the handoff of the chunk from the crashed pod to the new pod.
+
+In the case of network partitions between PPS and a pod, PPS will again stop receiving heartbeats and thus revoke the lease.  Pods are also programmed to crash if they can't reach PPS, causing new pods to be spawned in their place.  Thus, we again have a handoff of chunks from old, bad pods to new, functioning pods.
+
+## Finish jobs
+
+When a pod finishes processing a chunk, it sends a `FinishPod` request to PPS.  PPS then updates the relevant `Chunk` document to indicate that the chunk has been finished.
+
+When a job is first started, a `jobManager` goroutine is launched.  The `jobManager` subscribes to a changefeed that returns the statuses of the job's chunks.  When all chunks have been finished, the `jobManager` updates the state of the job (which is also stored as a separate document in Rethink) to indicate that the job has been finished.
+
+## Exercises
```

Comment:
OK I was thinking that it might help for people to understand the system by actively asking themselves those questions, but I have no problem removing it.